heidrun/nara_json.rb

```diff
@@ -0,0 +1,575 @@
+  # if record.find() did not match anything, but it's not clear how to achieve
+  # that.
+  RDF::Literal.new(nil)
+end
+
+# Return a string for sourceResource.contributor
+#
+# Use <contributorType>Most Recent</contributorType> if multiple
+# <contributor-display> values.
+# Reject <contributorType>Publisher</contributorType>.
+#
+#  <organizationalContributorArray> OR <personalContributorArray>
+#    <organizationalContributor> OR <personalContributor>
+#      <contributor>
+#        <termName>[VALUE]</termName>
+#      </contributor>
+#      <contributorType>
+#        <termName>[VALUE]</termName>
+#      </contributorType>
+#    </organizationalContributor> OR </personalContributor>
+#  </organizationalContributorArray> OR </personalContributorArray>
+#
+# @see #contributor_term_name
+# @return [String, RDF::Literal]
+def make_contributor(contributor_array)
+  contributor_term_name(contributor_array) do |contributors|
+    # Always reject 'Publisher' and use 'Most Recent' if more than one
+    contributors.select! do |c|
+      c if c['contributorType']['termName'] != 'Publisher'
+    end
+    if contributors.count > 1
+      contributors.select! do |c|
+        c if c['contributorType']['termName'] == 'Most Recent'
+      end
+    end
+  end
+end
+
+# Return a string for sourceResource.publisher
+#
+# **note these are contingent on the value of contributorType/termName
+# being ""Publisher""
+#
+# <organizationalContributorArray>
+#  <organizationalContributor>
+#  <contributor>
+#  <termName>[VALUE]</termName>
+#  </contributor>
+#  <contributorType>
+#  <termName>Publisher</termName>
+#  </contributorType>
+#  </organizationalContributor>
+#  </organizationalContributorArray>
+#
+#  <personalContributorArray>
+#  <personalContributor>
+#  <contributor>
+#  <termName>[VALUE]</termName>
+#  </contributor>
+#  <contributorType>
+#  <termName>Publisher</termName>
+#  </contributorType>
+#  </personalContributor>
+#  </personalContributorArray>
+#
+# @see #contributor_term_name
+# @return [String, RDF::Literal]
+def make_publisher(contributor_array)
+  contributor_term_name(contributor_array) do |contributors|
+    contributors.select! do |c|
+      c if c['contributorType']['termName'] == 'Publisher'
+    end
+  end
+end
+
+# Return string for sourceResource.identifier
+#
+# <variantControlNumberArray>
+#  <variantControlNumber>
+#  <number>[VALUE1]</number>
+#  <type>
+#  <termName>[VALUE2]</termName>
+#  </type>
+#  </variantControlNumber>
+#  </variantControlNumberArray>
+# [combine as:  VALUE2: VALUE1
+#
+# @param variant_control_num [Krikri::JsonParser::Value]
+# @return [String]
+def make_identifier(variant_control_num)
+  node = variant_control_num.node
+  # The 'type' property can be missing.
+  # See https://catalog.archives.gov/api/v1?naIds=65523&pretty=true&resultTypes=item,fileUnit&objects.object.@objectSortNum=1
+  type = (node.include? 'type') ? node['type']['termName'] : nil
+  [type, node['number']].compact.join ': '
```

Comment:
I'm not sure if this is more idiomatic, but I might do something like `"#{type ? type + ": " : nil}#{node['number']}"`. I don't have a strong feeling about this, and it'd probably benefit from someone else's comments. :)